### DISCUSSION TITLE
Building USWDS-based Apps on Next.js with SSR

### DISCUSSION THREAD

**Discussion by jbouder** [uswds/uswds#6107] [Q&A]
I'm curious to hear if anyone is effectively using USWDS on a Next.js app with SSR (or some other SSR framework). Based on my understanding and limited experience with Next.js, it appears as if it doesn't seem possible to effectively use USWDS on an SSR Next.js app...this being due to the fact that some USWDS components make use of the document and/or window objects.

So based on that, my assumption is that any pages that make use of USWDS components can't be server rendered, effectively eliminating the major reason for using Next.js. And my assumptions are based on current, production ready Next.js functionality. I'm guessing Partial Pre-Rendering might help with that.

Thanks!

**Comment by echappen** [comment]
Many (but not all) USWDS components require client-side interactivity. USWDS JS is not component-ized, it's imported wholesale which works against many Next.js paradigms. For our Next.js/USWDS [project](https://github.com/cloud-gov/cg-ui) (using USWDS 3x), we believed our best bet was to re-implement any USWDS JS in React, then decide from there whether a component could be server-rendered. Ad-Hoc's [uswds-react](https://github.com/trussworks/react-uswds) repo provided a good starting point for that.

USWDS [is moving to web components](https://github.com/uswds/uswds-proposals/blob/main/decisions/0001-use-web-components.md), which I believe are becoming increasingly compatible with SSR frameworks (still with kinks to work out though). I haven't tried that approach myself, so I'm curious how that progresses.

**Reply by jbouder** [reply]
Yeah, I'd like to avoid reproducing USWDS code in my own implementation, seems like it would be a pain to deal with updates from USWDS.

And in terms of Web Components, I believe they've mentioned that they are not focusing heavily on that at the moment with the new Web Components, so likely to still be an issue, until it becomes a priority.

Thanks for the insight!

**Comment by ryparker** [comment]
USWDS can be used on a server-rendered Next.js page. The browser-dependent JavaScript needs to run after mounting, but the page's HTML and USWDS styles can still be server-rendered.

Two clarifications to the earlier replies:

- USWDS exports individual JavaScript modules, such as `@uswds/uswds/js/usa-accordion`. Importing the entire auto-initializing bundle is not required.
- In Next.js, a Client Component can still have its initial HTML rendered on the server. Adding `'use client'` does not itself make the whole page client-rendered. See [Next.js's server and client component guidance](https://nextjs.org/docs/app/getting-started/server-and-client-components).

For an accordion, use a small Client Component wrapper with a DOM ref. Dynamically import its USWDS module inside `useEffect`, call `.on(ref.current)` after the markup mounts, then `.off()` with that same element during cleanup. Guard the asynchronous import so it does not initialize after unmount. Importing a browser-dependent module at the top of the file can still evaluate it during server rendering.

Keep React and USWDS from independently changing the same enhanced markup, and account for each component's cleanup behavior when content changes. This approach does not require rewriting the whole site in client-side React or waiting for a future rendering feature.
